### PR TITLE
App testing

### PR DESCRIPTION
**Pull request by vfilimo**
coverage

### CODE REVIEW

File: src/test/java/online/store/book/service/book/BookServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+@ExtendWith(MockitoExtension.class)
+class BookServiceImplTest {
+    @InjectMocks
```
Comment (MentorMateAcademy) [inline src/test/java/online/store/book/service/book/BookServiceImplTest.java]: inject mocks after mocks

File: src/test/java/online/store/book/service/book/BookServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    @Mock
+    private BookRepository bookRepository;
+    @Mock
+    private BookMapper bookMapper;
+    @Mock
+    private BookSpecificationBuilder specificationBuilder;
+    private CreateBookRequestDto bookRequestDto;
+    private Book book;
+    private BookDto bookDto;
+    private Pageable pageable;
```
Comment (MentorMateAcademy) [inline src/test/java/online/store/book/service/book/BookServiceImplTest.java]: let`s build this entities by private methods

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/test/java/online/store/book/controller/BookControllerIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+import org.springframework.jdbc.datasource.init.ScriptUtils;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerIntegrationTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/controller/BookControllerIntegrationTest.java]: Let's remove Integration. You don't have any other type of tests for Controller ) 
```suggestion
class BookControllerTest {
```

File: src/test/java/online/store/book/controller/BookControllerIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                .build();
+    }
+
+    @BeforeEach
+    void setUp(@Autowired DataSource dataSource) throws Exception {
+        teardown(dataSource);
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("db/books/add_books_for_Integration_Test.sql"));
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/controller/BookControllerIntegrationTest.java]: use a uniform naming style for files
```suggestion
                    new ClassPathResource("db/books/add-books-for-integration-test.sql"));
```

File: src/test/java/online/store/book/BookRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        assertEquals(1L, actualFirstCategory.getTotalElements());
+        assertEquals("John Doe", actualAuthor);
+
+        Page<Book> actualSecondCategory = bookRepository.findAllByCategoryId(2L, pageable);
+        String actualTitle = actualSecondCategory.stream()
+                .findFirst()
+                .get()
+                .getTitle();
+        assertEquals(2L, actualSecondCategory.getTotalElements());
+        assertEquals("Horror Stories", actualTitle);
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/BookRepositoryTest.java]: This is a rather strange check that relies on the sequence of elements in the collection. Why then do we validate the author and not the title of the book?

File: src/test/java/online/store/book/BookRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            "classpath:db/books/remove_books_from_db.sql",
+            "classpath:db/categories/remove_categories_from_db.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void findAllByCategoryId_searchByNotExistingCategory_shouldReturnEmpty() {
+        Pageable pageable = PageRequest.of(0, 10, Sort.by("id"));
+        Page<Book> bookPage = bookRepository.findAllByCategoryId(3L, pageable);
+        boolean isPresent = bookPage.stream()
+                .findFirst()
+                .isPresent();
+        assertFalse(isPresent);
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/BookRepositoryTest.java]: The same. Better get the content of the Page and check that size == 0

File: src/test/java/online/store/book/controller/BookControllerIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        BookDto expectedBookDto = new BookDto();
+        expectedBookDto.setAuthor(bookRequestDto.getAuthor());
+        expectedBookDto.setTitle(bookRequestDto.getTitle());
+        expectedBookDto.setIsbn(bookRequestDto.getIsbn());
+        expectedBookDto.setPrice(bookRequestDto.getPrice());
+        expectedBookDto.setCategoryIds(bookRequestDto.getCategoryIds());
+
+        String jsonRequest = objectMapper.writeValueAsString(bookRequestDto);
+
+        MvcResult result = mockMvc.perform(post("/books")
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/controller/BookControllerIntegrationTest.java]: this literal should be constant

File: src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            @Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders.webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @BeforeEach
+    void setUp(@Autowired DataSource dataSource, TestInfo testInfo) throws Exception {
+        if (!testInfo.getDisplayName().contains("Find a books by an existing category id")) {
+            teardown(dataSource);
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java]: Why do you need this check? If you need this logic for a specific test use it above the test

File: src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+                .andReturn();
+    }
+
+    @Test
+    @WithMockUser(username = "user", roles = {"User"})
+    @DisplayName("Delete category with a non-correct user role")
+    void deleteCategory_InValidRole_NotSuccess() throws Exception {
+        MvcResult result = mockMvc.perform(delete("/categories/1"))
+                .andExpect(status().isForbidden())
+                .andReturn();
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java]: what are you testing here? security? 
Do you need to test the filter chain in this class?

File: src/test/java/online/store/book/service/book/BookServiceImplTest.java (hunk 2 above)
Comment (liliia-ponomarenko) [reply src/test/java/online/store/book/service/book/BookServiceImplTest.java]: no need to take it to the class level, it's better for each of your tests to have its own object, so just call the private method of creating the object in the appropriate tests

File: src/test/java/online/store/book/service/category/CategoryServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    private CategoryMapper categoryMapper;
+    @Mock
+    private CategoryRepository categoryRepository;
+    @InjectMocks
+    private CategoryServiceImpl categoryService;
+    private CategoryDto categoryDto;
+    private CreateCategoryRequestDto createCategoryRequestDto;
+    private Category category;
+    private Pageable pageable;
+
```
Comment (liliia-ponomarenko) [inline src/test/java/online/store/book/service/category/CategoryServiceImplTest.java]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java (hunk 8 above)
Comment (vfilimo) [reply src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java]: I need this check to do the setup for all methods except one.

Comment (okuzan) [reply src/test/java/online/store/book/controller/CategoryControllerIntegrationTest.java]: This approach looks fragile to me - depends of display name wording. Let's use something more solid, try smth like this
image
https://www.baeldung.com/junit-filtering-tests

**Review by okuzan — CHANGES_REQUESTED** [review]
left a suggestion for test filtering

**Review by okuzan — APPROVED** [review]

**Pull request merged**